src/core/ext/transport/binder/wire_format/wire_writer.cc

```diff
@@ -30,60 +30,160 @@ namespace grpc_binder {
-      }
+  if (CanBeSentInOneTransaction(tx)) {
+    return RpcCallFastPath(tx);
+  }
+  // Slow path: the message data is too large to fit in one transaction.
+  int& seq = seq_num_[tx.GetTxCode()];
+  int original_flags = tx.GetFlags();
+  GPR_ASSERT(original_flags & kFlagMessageData);
+  absl::string_view data = tx.GetMessageData();
+  size_t ptr = 0;
+  while (ptr < data.size()) {
+    if (!WaitForAcknowledgement()) {
+      return absl::InternalError("Timeout waiting for acknowledgement");
     }
-    if (tx.GetFlags() & kFlagMessageData) {
-      RETURN_IF_ERROR(parcel->WriteByteArrayWithLength(tx.GetMessageData()));
+    RETURN_IF_ERROR(binder_->PrepareTransaction());
+    WritableParcel* parcel = binder_->GetWritableParcel();
+    size_t size = std::min(static_cast<size_t>(kBlockSize), data.size() - ptr);
+    int flags = GetFlagForCurrentTransaction(
+        original_flags, /*sent_length=*/ptr,
+        /*remaining_length=*/data.size() - ptr - size);
+    RETURN_IF_ERROR(parcel->WriteInt32(flags));
+    RETURN_IF_ERROR(parcel->WriteInt32(seq++));
+    if (flags & kFlagPrefix) {
+      RETURN_IF_ERROR(WriteInitialMetadata(tx, parcel));
     }
-    if (tx.GetFlags() & kFlagSuffix) {
-      if (tx.IsServer()) {
-        if (tx.GetFlags() & kFlagStatusDescription) {
-          RETURN_IF_ERROR(parcel->WriteString(tx.GetStatusDesc()));
-        }
-        RETURN_IF_ERROR(parcel->WriteInt32(tx.GetSuffixMetadata().size()));
-        for (const auto& md : tx.GetSuffixMetadata()) {
-          RETURN_IF_ERROR(parcel->WriteByteArrayWithLength(md.first));
-          RETURN_IF_ERROR(parcel->WriteByteArrayWithLength(md.second));
-        }
-      } else {
-        // client suffix currently is always empty according to the wireformat
-        if (!tx.GetSuffixMetadata().empty()) {
-          gpr_log(GPR_ERROR, "Got non-empty suffix metadata from client.");
-        }
-      }
+    RETURN_IF_ERROR(parcel->WriteByteArrayWithLength(data.substr(ptr, size)));
+    if (flags & kFlagSuffix) {
+      RETURN_IF_ERROR(WriteTrailingMetadata(tx, parcel));
     }
+    RETURN_IF_ERROR(binder_->Transact(BinderTransportTxCode(tx.GetTxCode())));
+    ptr += size;
+    // TODO(waynetu): This should be parcel->GetDataSize().
+    num_outgoing_bytes_ += size;
   }
-  // FIXME(waynetu): Construct BinderTransportTxCode from an arbitrary integer
-  // is an undefined behavior.
-  return binder_->Transact(BinderTransportTxCode(tx.GetTxCode()));
+  return absl::OkStatus();
 }
 
-absl::Status WireWriterImpl::Ack(int64_t num_bytes) {
+absl::Status WireWriterImpl::SendAck(int64_t num_bytes) {
   grpc_core::MutexLock lock(&mu_);
   RETURN_IF_ERROR(binder_->PrepareTransaction());
   WritableParcel* parcel = binder_->GetWritableParcel();
   RETURN_IF_ERROR(parcel->WriteInt64(num_bytes));
   return binder_->Transact(BinderTransportTxCode::ACKNOWLEDGE_BYTES);
 }
 
+void WireWriterImpl::RecvAck(int64_t num_bytes) {
+  grpc_core::MutexLock lock(&mu_);
```

Comment:
there is no deadlock?

Is it because `cv_.WaitWithDeadline` allows us to obtain `mu_` here?

Reply:
Waiting on a conditional variable will release the mutex IIRC.

Reply:
Ok. (I know it is true in boost, just want to confirm if that's also the case in abseil)